jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/WebSocketDemander.java

```diff
@@ -37,15 +37,17 @@
  * implements the {@link DemandChain} interface. The methods of {@link DemandChain} from the {@link Extension}
  * must be forwarded to this flusher.</p>
  */
-public abstract class DemandingFlusher extends IteratingCallback implements DemandChain
+public abstract class WebSocketDemander extends IteratingCallback implements DemandChain
 {
     private static final Throwable SENTINEL_CLOSE_EXCEPTION = new StaticException("Closed");
 
+    private final AutoLock _lock = new AutoLock();
     private final IncomingFrames _emitFrame;
-    private final AtomicBoolean _demand = new AtomicBoolean();
-    private final AtomicReference<Throwable> _failure = new AtomicReference<>();
-    private DemandChain _nextDemand;
 
+    private boolean _upstreamDemand = false;
+    private boolean _downstreamDemand = false;
```

Comment:
probably a little comment to clarify which direction upstream and downstream are.